foursquare/finatra
# patch finatra to use shaded jackson

There are a handful of steps involved here:

 0. add our shaded core jackson-fatjar and jackson-module-scala jars
    to the repo's top-level lib directory (they are unmanaged
    dependencies in sbt's eyes)
 1. update build configuration to ban upstream jackson jars
 2. update jackson imports to use their shaded variants
 3. compile
 4. once everything compiles, comment out the upstream jackson
    excludes
 5. run tests
 6. uncomment the upstream jackson excludes
 7. bump the patch version and publish the shaded finatra jars
 8. test using the shaded jars locally
 9. upload the jars, javadocs, sources, and poms to artifactory. You
    won't need all of them -- I recommend bumping the patch version
    in foursquare.web first, running ivy resolve, and uploading
    missing artifacts until it succeeds
 10. make a new branch on our github fork off the upstream release
    tag (call it something like 20.1.0-fs), open a pr against that
    for review, then merge your changes and push a new git tag for
    the patch release (eg. finatra-20.1.0-fs0)

Code-wise, this is a dead simple namespace refactor. Unfortunately,
most of the complexity involved comes at runtime in the form of
manually validating the classpath is properly constructed. The repo's
test suite certainly helps and is a good place to start, however some
other checks to make:

 - in sbt, run `show test:fullClasspath` and verify it shows no
   versions of jackson other than our shaded fat jar
 - as an added sanity check, always unzip our shaded jackson jars and
   make sure they doesn't include unshaded jackson class files

Assuming you have done your due diligence on all of the above, your
newly shaded jars should be good to go!

There are some open questions on how sbt treats its test classpath.
Any tests using shaded jackson fail with classloading errors when
`fork in Test := true` is set but pass when it isn't, while there are
a small handful of tests using the scala-parser-combinators library
that fail only when `fork in Test := true` _isn't_ set (also
reproducible on upstream finatra!). While I would love answers, I'm
reasonably convinced these are sbt gremlins unrelated to the changes
here and have given up trying to find a fix.
